ceno_rt/src/allocator.rs

```diff
@@ -4,26 +4,34 @@
 use core::alloc::{GlobalAlloc, Layout};
 
 struct SimpleAllocator {
-    next_alloc: usize,
+    next_alloc: *mut u8,
 }
 
+extern "C" {
+    // The address of this variable is the start of the heap (growing upwards).
+    static mut _sheap: u8;
+}
+
+#[global_allocator]
+static mut HEAP: SimpleAllocator = SimpleAllocator {
+    next_alloc: &raw mut _sheap,
+};
+
 unsafe impl GlobalAlloc for SimpleAllocator {
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
         // SAFETY: Single threaded, so nothing else can touch this while we're working.
         let mut heap_pos = HEAP.next_alloc;
 
         let align = layout.align();
         // `Layout` contract forbids making a `Layout` with align=0, or align not power of 2.
-        // So we can safely use subtraction and a mask to ensure alignment without worrying about UB.
-        let offset = heap_pos & (align - 1);
-        if offset != 0 {
-            heap_pos = heap_pos.strict_add(align.strict_sub(offset));
-        }
+        core::hint::assert_unchecked(align.is_power_of_two());
+        core::hint::assert_unchecked(align != 0);
+        heap_pos = heap_pos.add(heap_pos.align_offset(align));
```

Comment:
We use the proper Rust functions for working with pointers now.  That makes both the type checker happier (when working with pointers instead of `usize`), and hopefully also clues in the optimiser a bit more.

But the latter would just be a welcome side effect, if it does happen.